src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -431,6 +436,21 @@ public override IEnumerable<TResult> Where(Func<TResult, bool> predicate)
             {
                 return new WhereEnumerableIterator<TResult>(this, predicate);
             }
+
+            public override TResult[] ToArray()
+            {
+                if (_predicate != null)
+                {
+                    return base.ToArray();
+                }
+
+                var results = new TResult[_source.Length];
+                for (int i = 0; i < _source.Length; i++)
+                {
```

Comment:
result.Length   could result in a better range check here. Since result is a local, it is easier to prove its Length cannot change.